src/cleanvision/videolab.py

```diff
@@ -0,0 +1,260 @@
+"""Videolab is an extension of Imagelab for finding issues in a video dataset."""
```

Comment:
amazing stuff!  

Could you comment:

A minimal example script of how to use this new Videolab and what the outputs look like?

A comment with comprehensive list of limitations of this code?
Eg: 
- What video file-type requirements are there. 
- What happens if some videos are really long some really short.
- any efficiency bottlenecks where the code feels slow
- edge cases where the results might not be good or code might crash.

Reply:
# Limitations
Exhaustive survey of the limitations of the current implementation for `Videolab` for *pull request #231*.

## General
The following is a brief *summary* of the general limitations of this pull request. For a more detailed (i.e. *low level*) analysis, see the **Specifics** section.

+ **NO** multiprocessing or multithreading used
+ **NO** default arguments used
+ **NO** parameter/error checking used
+ **NO** progress bars used (i.e. `tqdm`)
+ **NOT** all methods are implemented/overrided in `Videolab` class
+ **ONLY** these *video formats* are searched: https://github.com/lafemmephile/cleanvision/blob/43584bcb5554d52236d657d1fb9e7f1664eaf101/src/cleanvision/videolab.py#L12

*Why?* From the discussion in #215, I decided it was best to minimize any *excess* decision making, since the `cleanvision` maintainers will have their own *standards* and *needs* for their library.

## Specifics
Here we will dig into the actual *code* and discuss the *select practical issues* that are currently present.

### Frame Sampling
Let us begin by looking at the *frame sampling* class aptly named `FrameSampler`:
https://github.com/lafemmephile/cleanvision/blob/43584bcb5554d52236d657d1fb9e7f1664eaf101/src/cleanvision/videolab.py#L15-L52

#### Creating Frame Sample Directory: Collisions
One of the first real issues is in the *strategy* for naming the directories corresponding to the *frame samples* from a specific video file:
https://github.com/lafemmephile/cleanvision/blob/43584bcb5554d52236d657d1fb9e7f1664eaf101/src/cleanvision/videolab.py#L22-L29

The current implementation uses the *user supplied* `output_dir` and the actual *file name* of the video. This may lead to situations where *collisions* occur: two video files have the same name, but different paths, and will map to the same *frame sample sub directory*. For example see the following diagram:
```console
/home/foo/cats/videos/001.mp4
               |
                \----> sample_frames/001.mp4/ <----\
                                                    |
                                        /home/foo/downloads/misc/memes/001.mp4
```

One way around this is by using some kind of *hash*, either on the *absolute file path* or the *video file contents*:
```console
/home/foo/cats/videos/001.mp4
               |
             HASH()
               |
                \----> sample_frames/9asd09kl23kl/001.mp4/

                sample_frames/jd82lkxja983/001.mp4/ <----\
                                                          |
                                                        HASH()
                                                          |
                                        /home/foo/downloads/misc/memes/001.mp4
```

Certainly there are multiple ways this could be done. The above shows one potential strategy.

### Videolab Class
Now let us dive into the actual `Videolab` class:
https://github.com/lafemmephile/cleanvision/blob/43584bcb5554d52236d657d1fb9e7f1664eaf101/src/cleanvision/videolab.py#L55-L260

#### Constructor
One of the first real changes or *issues* with the `Videolab` class in comparison to `Imagelab` is the `__init__()` method:
+ `Videolab`: https://github.com/lafemmephile/cleanvision/blob/43584bcb5554d52236d657d1fb9e7f1664eaf101/src/cleanvision/videolab.py#L58-L64
+ `Imagelab`: https://github.com/lafemmephile/cleanvision/blob/43584bcb5554d52236d657d1fb9e7f1664eaf101/src/cleanvision/imagelab.py#L118-L148

As you can see in the comparison above, `Videolab` does not use the `Imagelab` *constructor* but instead simply takes in a *directory path* as a `str` and stores it as a `pathlib.Path` object. This means that currently `Videolab` does not accept the other *dataset* parameters that `Imagelab` does. The reason why this was chosen will be reviewed in the **Finding Issues** section.

#### Frame Sampling: Single Process/Thread
One of the obvious performance issues will be in the actual *frame sampling*:
https://github.com/lafemmephile/cleanvision/blob/43584bcb5554d52236d657d1fb9e7f1664eaf101/src/cleanvision/videolab.py#L73-L81

Whereas the original [VideoSampler](https://github.com/LemurPwned/video-sampler) written by @LemurPwned used multiprocessing, I chose to write a *minimal* version that does not. That way the *performance strategy* for the frame sampler can be dealt with separately in a manner that fits the maintainer's needs.

#### Aggregating Image Issues
One of the *unique* tasks that `Videolab` tackles is the *aggregation* of issues found
by `Imagelab`:
https://github.com/lafemmephile/cleanvision/blob/43584bcb5554d52236d657d1fb9e7f1664eaf101/src/cleanvision/videolab.py#L108-L139

There may be some value in considering the *performance* of the *aggregation algorithm* which essentially store a `pandas.DataFrame` of the *mean* of all the *video frames* for a specific video in a `list`, which is finally *concatenated* with the `pandas.concat` function. This may or may not be the *optimal* way to aggregate issues.

Another point worth considering is what *sample frame* to use as the **"image"** that is representative of the video. Currently this line calculates which *sample frame* to use as a new index in the new aggregate issue:
https://github.com/lafemmephile/cleanvision/blob/43584bcb5554d52236d657d1fb9e7f1664eaf101/src/cleanvision/videolab.py#L121-L122

The current strategy is to use *approximately* the *middle frame* to represent the video in the new *aggregate issues* `pandas.DataFrame`.

#### Finding Issues
The `cleanvision` maintainers may choose a different approach to how I have implemented `find_issues()`:
https://github.com/lafemmephile/cleanvision/blob/43584bcb5554d52236d657d1fb9e7f1664eaf101/src/cleanvision/videolab.py#L162-L182

Specifically they will notice that opted to execute the *frame sampling* step in the `find_issues` method, and then subsequently call the *constructor* for `Imagelab`, and then call the `Imagelab.find_issues` parent method.

Reply:
> A comment with comprehensive list of limitations of this code? Eg:
> 
>     * What video file-type requirements are there.
> 
>     * What happens if some videos are really long some really short.
> 
>     * any efficiency bottlenecks where the code feels slow
> 
>     * edge cases where the results might not be good or code might crash.

@jwmueller I will now look at developing some *tests* and *performance experiments* in this [repo](https://github.com/lafemmephile/videolab) to attempt to rigorously answer these questions.

Reply:
# Performance
Using the *Videolab demo* found [here](https://github.com/lafemmephile/videolab), the following `scalene` profiles were generated.

## Reduced Profile
![Screenshot from 2023-10-23 14-41-31](https://github.com/cleanlab/cleanvision/assets/126025474/dae57fbe-b475-42b5-a754-1f4c79de364d)

## Full Profile
![Screenshot 2023-10-23 at 14-46-53 Screenshot](https://github.com/cleanlab/cleanvision/assets/126025474/bb9f7890-688b-4d04-aead-3cf1229ff7ec)

Reply:
@jwmueller I've done some *preliminary* performance profiling using `scalene` (see above comment). Next I want to scale it up to a **MUCH LARGER** dataset (not just the measly 6 videos from the *demo) and do some *performance* profiling.

Reply:
# Performance: Videolab Demo 5K
Using the *Videolab demo* found [here](https://github.com/lafemmephile/videolab/blob/59be68966873057900936ecf3d90514cb9814a24/notebooks/videolab_demo.ipynb) (and available in *Google Colab* [here](https://colab.research.google.com/github/lafemmephile/videolab/blob/master/notebooks/videolab_demo.ipynb)), the following `scalene` profiles were generated. In the course of the *performance profiling* a total of **4999 videos** are downloaded (from the [WebVid-10M dataset](https://maxbain.com/webvid-dataset/)), sampled, and analyzed by `Imagelab`.

# Reduced Profile
```
scalene --reduced-profile --cli --html --outfile reduced_profile.html videolab_demo.py
```
![Screenshot 2023-10-23 at 19-39-53 Screenshot](https://github.com/cleanlab/cleanvision/assets/126025474/6ae14a60-9b53-441c-a8fe-e9a5975a260b)

Reply:
@jwmueller From generating these `scalene` profiles, I can honestly say that the `FrameSampler` is horribly slow (something @LemurPwned and @smttsp were well aware of). For example, running the above `Videolab` on the *5k* sub-sample of the [WebVid-10M dataset](https://maxbain.com/webvid-dataset/)  took ~40 minutes to sample frames from 4999 videos, whereas `Imagelab` was able to process about *88k* images in ~10 minutes.

Hence, performance appears to be most affected by the *frame sampling strategy* than any other step in `Videolab`. Whether in this *pull request* or another separate issue, it will require a far more *optimized* and *intelligent* design than the simple `FrameSampler` that I have implemented (something closer to @LemurPwned's [VideoSampler](https://github.com/LemurPwned/video-sampler).

Reply:
Great callout, ya replacing the FrameSampler will certainly need to be improved before we publicly announce this module to users.  But I don't think it is critical to this PR being merged and could be addressed in a future follow-up PR.

Thanks for the comprehensive benchmarking/profiling, very helpful!  Our team will get to the code review as soon as we're able to

Reply:
Great job, if you need me to optimise this, let me know, there are some avenues to explore here.

Reply:
@LemurPwned Thank you. It was a team effort by all of you. I just did what I could to *"lay the foundation"*. Yes definitely after this initial PR (like @jwmueller suggested) we can really optimize the *frame sampling*, and your and @smttsp's expertise in this matter are invaluable.

Reply:
@lafemmephile Do you want to make one of your notebooks here:
https://github.com/lafemmephile/videolab/tree/master/notebooks

the official quickstart tutorial for Videolab? 

Here's a GH issue detailing what we need for our quick-start tutorial:
https://github.com/cleanlab/cleanvision/issues/239

Should be a simple PR of one of your existing notebooks, which you can make separately from the Videolab source code PR in parallel, if you're interested!

Reply:
Definitely! I will work on this as well!